src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -2491,12 +2507,37 @@ public override BoundNode VisitConditionalAccess(BoundConditionalAccess node)
                     SetUnreachable();
                 }
 
-                VisitRvalue(node.AccessExpression);
-                Join(ref this.State, ref savedState);
+                // We want to preserve stateWhenNotNull from accesses in the same "chain":
+                // a?.b(out x)?.c(out y); // expected to preserve stateWhenNotNull from both ?.b(out x) and ?.c(out y)
+                // but not accesses in nested expressions:
+                // a?.b(out x, c?.d(out y)); // expected to preserve stateWhenNotNull from a?.b(out x, ...) but not from c?.d(out y)
+                BoundExpression expr = node.AccessExpression;
+                while (expr is BoundConditionalAccess innerCondAccess)
+                {
+                    // we assume that non-conditional accesses can never contain conditional accesses from the same "chain".
+                    // that is, we never have to dig through non-conditional accesses to find and handle conditional accesses.
+                    VisitRvalue(innerCondAccess.Receiver);
+                    expr = innerCondAccess.AccessExpression;
+                }
```

Comment:
From our offline discussion, we need to not only maintain the "state when everything went well" (stateWhenNotNull), but also the worst case state (which factors all the possibilities, such as the chain partially executing).
Your example `a?.b(x = null)?.c(x = new object())` illustrates the problem.

Reply:
Done, thanks.